### DISCUSSION TITLE
Is it possible to lazy calculate dag start_date on first run

### DISCUSSION THREAD

**Discussion by argibbs** [apache/airflow#23594] [Q&A] [answered]
Hi guys,

I'm on airflow 2.2.5 (will be 2.3.1 when it's available)

In our set up, we have dags with external dependencies where one dag has external task sensors that depend on dag run _from the previous day_ in a second dag.

e.g. The dag run for `Dag_OneDayLag` on Tuesday has an external task sensor on the dag run for `Dag_Live` for Monday.

This all works fine, but when we deploy the dags for the first time, the first dag run for `Dag_OneDayLag` hangs, because it's looking for a dag run in `Dag_Live` that doesn't exist. This is solvable by backfilling `Dag_Live` by one day, but it's clunky and manual.

Also, Because Of Reasons (TM), these dags have `catchup=False`. If they get recreated, I don't want these dags to start regenerating from a year ago or whenever they were first created. However, `catchup=False` is actually not ideal, I'd prefer to be able to set catchup=True, but if I delete and recreate one of these dags, I really want it to restart from today (or in the case of Dag_Live, yesterday). I could also simply track and update the start_date for the dags, but it's again clunky and manual, _and_ requires republishing the dag definitions each time I want to recreate them.

What would solve all of this for me is if there was a way to set the start_date based on the current date on the first run. 

I could of course do the dumb thing and just always work out what today is, and then set `start_date` to be either `today` or something like `today.add(days=-1)` depending on which dag it is. The catch is that this means that `start_date` is an ever shifting goalpost and I seem to recall reading somewhere that this is a Bad Idea.

So, questions:
1. Is it ok to keep changing the start_date even if there are runs prior to that date?
2. Alternatively is there a way to check if the dag is "new" and thus determine a fixed start_date relative to today if it's new, or maintain the existing start_date if it's been run before?

Hopefully the above makes sense. Thanks!

**Comment by potiuk** [comment, marked as answer]
There is VERY APT discussion about it in the devlist :) https://lists.apache.org/thread/hyctljm0o1odn1bntp4fcgok0d0433ov 

In short. Yes. It makes sense. And we might do just that.

**Reply by argibbs** [reply]
Hah, you're right that is uncannily timed. Yes, I would definitely make use of such a feature.

**Comment by malthe** [comment]
Assuming a cron-based interval, perhaps `start_date` should default to the current time minus the data interval length.

This seems timetable-specific. Could we add a method to the timetable which would return a default `start_date` that makes sense for the specific timetable as the equivalent of basically "today" – ?

**Reply by argibbs** [reply]
Mmmm... but the timetable is not really a property of the dag; if what you're saying is viable, then is there any reason why I can't just do option 1 in my original question? i.e. always be setting start date to yesterday, even if the dag was first created days / weeks / months ago?

(Aside: I'm actually trying this as a stop-gap, tho too soon to be sure it's working properly)

As mooted in the dev-list discussion, it seems like the question is over whether `start_date` even needs to be set in most cases?

**Reply by malthe** [reply]
The timetable _is_ a [property of the dag](https://github.com/apache/airflow/blob/cfa95af7e83b067787d8d6596caa3bc97f4b25bd/airflow/models/dag.py#L313).

The important method is in the timetable interface:
```python
def next_dagrun_info(
    self,
    *,
    last_automated_data_interval: Optional[DataInterval],
    restriction: TimeRestriction,
) -> Optional[DagRunInfo]
```
And `restriction.earliest` is computed based on the `start_date` of the tasks in the DAG.

But it's a chicken and egg thing because there is no last automated DAG run if there is no `start_date` and so forth.

**Reply by argibbs** [reply]
> The timetable is a property of the dag

Yeah sorry I was tired and didn't express myself properly. I was trying to say start date not currently a property of the timetable. And as you then point out (you're better at this than me) you end up with a chicken and egg; thus I'm left wondering if it's ok to simply keep redefining start_date all the time.

Having had a bit more time to thing about it, I'm considering a stop gap where I do something like:
```
if len(dag_runs) == 0:
   # compute a dynamic start date
   # save it using Variable.set with a name specific to the dag
else:
  # read start_date using Variable.get
```

This seems to me like it'd tick all the boxes: if the dag is deleted, the start date will reset, but so long as the dag is running, the start date is fixed.

I've no idea if this will work / be performant / cause experts' jaws to drop in horror, but I might give it a try.

**Reply by argibbs** [reply]
Ok, so I've had a stab at this. The core of the code is below, seems to work well (there's a function or two I call but haven't pasted, am sure you could write your own, person-who-is-reading-this)

```python
def _write_start_date(*, variable_name: str, start_date: pendulum.DateTime):
    """Writes a dict of values suitable for passing to the pendulum.datetime() call that would reconstruct the given
    date time"""
    date_ctr_args: dict[str, Union[int, str]] = {"year": start_date.year, "month": start_date.month, "day": start_date.day, "tz": start_date.tz.name}
    Variable.set(key=variable_name, value=date_ctr_args, description="Dynamically generated start date for newly created dag", serialize_json=True)


def get_lazy_created_sod_start_date(*, dag_id: str, tz_name: str, start_days_ago: int = DEFAULT_START_DATE_OFFSET) -> pendulum.DateTime:
    variable_name: str = f"DynamicStartDate_{dag_id}"

    # if no dag runs, save the date as the new start date
    dag_runs: list[DagRun] = DagRun.find(dag_id)
    if not dag_runs:
        new_dynamic_sod = get_t_plus_n_sod(tz_name, start_days_ago)
        _write_start_date(variable_name=variable_name, start_date=new_dynamic_sod)
        return new_dynamic_sod

    # dag has run previously, use the value from the db
    date_ctr_args: dict[str, Union[int, str]] = Variable.get(key=variable_name, default_var=None, deserialize_json=True)
    if date_ctr_args is None:
        # this shouldn't happen, but if it did we don't want to 'splode
        logger.warning(f"Found {len(dag_runs)} previous dag runs for {dag_id}, but there was no persisted start date in the variables")

        first_known_start_date = get_sod_in_tz(tz_name=tz_name, dt_to_trunc=pendulum.instance(dag_runs[0].start_date))
        _write_start_date(variable_name=variable_name, start_date=first_known_start_date)
        return first_known_start_date

    return pendulum.datetime(**date_ctr_args).in_tz(tz_name)
```

My variables then look like this:
![image](https://user-images.githubusercontent.com/4406514/168092877-487aa610-48ae-4a4e-a1b0-f7c38ec3dd70.png)
